proposals/data-classes.md

```diff
@@ -0,0 +1,241 @@
+
+There is also no way to create a copy of a data structure with readonly
+fields with only one item changed. A new type must be constructed manually.
+
+
+## Proposal
+
+To resolve many of these issues, I propse a new modifier for classes and structs: `data`.
+`data` classes or structs are meant to satisfy the goals listed above by doing the
+following things:
+
+1. Automatically generating `Equals`, `GetHashCode`, `ToString`, `==`, `!=`, and `IEquatable<T>`
+   based on the member data of the type.
+1. Allow object initializers to also initialize readonly members.
+
+### Equality
+
+First, the generation of equality support. Data members are only public
+fields and auto-properties. This allows data classes to have private
+implementation details for things like caching without giving up simple
+equality semantics. There are a few places this could be problematic. For
+instance, only auto-properties are considered data members by default, but
+it's not uncommon to have some simple validation included in a property
+getter that does not meaningfully change the semantics, e.g.
+
+```C#
+{
+    ...
+    private int _field;
+    public int Field
+    {
+        get
+        {
+            Debug.Assert(_field >= 0);
+            return _field;
+        }
+    }
+}
+```
+
+To support these cases and provide an easy escape hatch, I propose a
+new attribute, `DataMemberAttribute` with a boolean flag argument on the
+constructor. This allows users to override the normal behavior and include
+or exclude extra members in equality. The previous example would now read:
+
+```C#
+{
+    ...
+    private int _field;
+
+    [DataMember(true)]
+    public int Field
+    {
+        get
+        {
+            Debug.Assert(_field >= 0);
+            return _field;
+        }
+    }
+}
+```
+
+Equality itself would be defined in terms of its data members. A `data` type
+is equal to another `data` type when there is an implicit conversion between
+the target type and the source type and each of the corresponding members
+are equal. The members are compared by `==` if it is available. Otherwise,
+the method `Equals` is tried according to overload resolution rules (st. an
+`Equals` method with an identity conversion to the target type is preferred
+over the virtual `Equals(object)` method).
+
+Note that the equality defined in this manner is not symmetric in the case
+of record subtyping. This topic will be addressed in more detail later, but
+it's important to note for equality that `a.Equals(b)` may be true if `B`
+is a subtype of `A` (and therefore has an implicit conversion) and all of
+the shared members between `A` and `B` are equal. In contrast, `A` would
+likely not have an implicit conversion to `B`, so they would not be
+considered equal.
+
+`GetHashCode` would be implemented by calling `GetHashCode` on each of
+the data members and these would be combined using a symmetric operation
+(like `xor`) to prevent ordering from affecting the results.
```

Comment:
I think the primary result of this would be a bad hash code.

Trying to pretend that two different data classes are the same if they have the same members sounds like a bad idea to me.

Reply:
This sounds like you're restating the previous comment. If `Equals` is symmetric, do you have any other objection?

Reply:
@agocke No, this is a different issue. The problem with using a symmetric operation for combining hash codes is that it means that e.g. `new Point { X = 1, Y = 0 }.GetHashCode()` returns the same value as `new Point { X = 0, Y = 1 }.GetHashCode()`.

I believe that in some data sets, this could increase the number of collisions by a lot.

The comment about pretending that types are the same is because I assumed the reason for symmetry is that you wanted e.g. `new Contoso.Person { FirstName = "Andy", LastName = "Gocke" }.GetHashCode()` to be the same as `new AdventureWorks.Person { LastName = "Gocke", FirstName = "Andy" }.GetHashCode()`.

Sorry I was unclear.

Reply:
Ah, in that case the intention is: yes, fields are deliberately unordered. However, there is no structural equality -- the types must be nominally compatible for two types to be considered equal. It sounds like we're also centering around subtypes not being equal to supertypes, so that would imply the types are equal as well.

Reply:
@agocke In that case, I think the order of fields should matter specifically for the purpose of computing the hash code. Hash codes should not be preserved across executions, so I think it's fine if changing the order of fields changes the hash code.

(Even stronger example is .Net Core 2.1 `HashCode`. It uses randomization, which means hash codes produced by it change even if the types themselves don't. If that behavior is acceptable, then I think depending on the order of fields should be also acceptable, even if they are otherwise unordered.)